src/App.tsx

```diff
@@ -1,26 +1,506 @@
+
+      if (!todo) {
+        return;
+      }
+
+      const updatedTitle = { title: title.trim() };
+
+      updateTodo(id, updatedTitle)
+        .then(() => {
+          setTodos(currentTodos =>
+            currentTodos.map(item =>
+              item.id === id ? { ...item, title: updatedTitle.title } : item,
+            ),
+          );
+
+          setAllTodos(currentTodos => {
+            return currentTodos.map(item =>
+              item.id === id ? { ...item, title: updatedTitle.title } : item,
+            );
+          });
+        })
+        .catch(() => {
+          setErrorMessage('Unable to update a todo');
+        })
+        .finally(() => {
+          setIsUpdating(null);
+        });
+    }
+
+    if (title.trim() === '') {
+      setDeleteTodoId(id);
+    }
+  }
+
+  function handleUpdateTodo(id: number) {
+    const todo = todos.find(currentTodo => currentTodo.id === id);
+
+    if (!todo) {
+      return;
+    }
+
+    const updatedStatus = { completed: !todo.completed };
+
+    if (updatedStatus) {
+      setIsUpdating(id);
+      setErrorMessage(null);
+
+      updateTodo(id, updatedStatus)
+        .then(() => {
+          setTodos(currentTodos =>
+            currentTodos.map(item =>
+              item.id === id
+                ? { ...item, completed: updatedStatus.completed }
+                : item,
+            ),
+          );
+
+          setAllTodos(currentTodos =>
+            currentTodos.map(item =>
+              item.id === id
+                ? { ...item, completed: updatedStatus.completed }
+                : item,
+            ),
+          );
+
+          handleFilter(sortBy);
+
+          if (updatedStatus.completed) {
+            setItemsLeft(currentItemsLeft => currentItemsLeft - 1);
+          } else {
+            setItemsLeft(currentItemsLeft => currentItemsLeft + 1);
+          }
+        })
+        .catch(() => {
+          setErrorMessage('Unable to update a todo');
+        })
+        .finally(() => {
+          setIsUpdating(null);
+        });
+    }
   }
 
+  function handleUpdateAllTodos() {
+    const hasUncompleteTodo = todos.findIndex(todo => !todo.completed);
+
+    if (hasUncompleteTodo >= 0) {
+      const updatedStatus = { completed: true };
+
+      const updateTodos = async () => {
+        const updatedTodos: Promise<number | null>[] = todos.map(async todo => {
+          if (!todo.completed) {
+            setIsUpdating(todo.id);
+            setErrorMessage(null);
+
+            try {
+              await updateTodo(todo.id, updatedStatus);
+
+              setItemsLeft(currentItemsLeft => currentItemsLeft - 1);
+
+              return todo.id;
+            } catch {
+              setErrorMessage('Unable to update a todo');
+
+              return null;
+            } finally {
+              setIsUpdating(null);
+            }
+          }
+
+          return null;
+        });
+
+        const updatedTodo = await Promise.all(updatedTodos);
```

Comment:
```suggestion
        const updatedTodos = await Promise.all(updatedTodos);
```
But Promise.all will be rejected if at least one todo update was failed. You should handle result for each todo separately